lib/src/database/indexeddb_box.dart

```diff
@@ -0,0 +1,194 @@
+import 'dart:async';
+import 'dart:html';
+import 'dart:indexed_db';
+
+/// Key-Value store abstraction over IndexedDB so that the sdk database can use
+/// a single interface for all platforms. API is inspired by Hive.
+class BoxCollection {
+  final Database _db;
+  final Set<String> boxNames;
+
+  BoxCollection(this._db, this.boxNames);
+
+  static Future<BoxCollection> open(
+    String name,
+    Set<String> boxNames, {
+    int version = 1,
+    Object? sqfliteDatabase,
+    IdbFactory? idbFactory,
+  }) async {
+    idbFactory ??= window.indexedDB!;
+    final db = await idbFactory.open(name, version: 1,
+        onUpgradeNeeded: (VersionChangeEvent event) {
+      final db = event.target.result;
+      for (final name in boxNames) {
+        db.createObjectStore(name, autoIncrement: true);
+      }
+    });
+    return BoxCollection(db, boxNames);
+  }
+
+  Box<V> openBox<V>(String name) {
+    if (!boxNames.contains(name)) {
+      throw ('Box with name $name is not in the known box names of this collection.');
+    }
+    return Box<V>(name, this);
+  }
+
+  List<Future<void> Function(Transaction txn)>? _txnCache;
+
+  Future<void> transaction(
+    Future<void> Function() action, {
+    List<String>? boxNames,
+    bool readOnly = false,
+  }) async {
+    boxNames ??= _db.objectStoreNames!.toList();
+    _txnCache = [];
+    await action();
+    final cache = List<Future<void> Function(Transaction txn)>.from(_txnCache!);
+    _txnCache = null;
+    if (cache.isEmpty) return;
+    final txn = _db.transaction(boxNames, readOnly ? 'readonly' : 'readwrite');
+    for (final fun in cache) {
+      unawaited(fun(txn));
+    }
+    await txn.completed;
+    return;
+  }
+
+  Future<void> clear() async {
+    for (final name in boxNames) {
+      _db.deleteObjectStore(name);
+    }
+  }
+
+  Future<void> close() async => _db.close();
+}
+
+class Box<V> {
+  final String name;
+  final BoxCollection boxCollection;
+  final Map<String, V?> _cache = {};
+  Set<String>? _cachedKeys;
+  bool get _keysCached => _cachedKeys != null;
```

Comment:
_cachedKeys is only used to make sure that if you fetch all keys from a box, you do not need to have an expensive read operation twice. There is no other usage for this at the moment (not sure if it is even worth the effort). Once the keys are cached, they need to be updated when changed in put and delete* so that the cache does not become outdated.

Do you think it makes sense to add this as a comment there?

Reply:
Imo yes. This would help me to not "fix" the "bug" in the future, that it is only updated sometimes, which could introduce other bugs. Basically I would like to see documented:

- Is the cache partial or complete. As in, does it contain all keys or only some recently read keys. That is important if you want to avoid a db read in methods that read all values for example.
- Is it lazy or eager? Basically is the cache initialized on first read, only in specific cases or when the box is opened.
- If it is updated or dropped on writes. I think in this case this means it gets updated.